File: 3525-find-x-value-of-array-ii/solution.py
from typing import List

class SegmentTree:
    def __init__(self, nums: List[int], k: int):
        self.n = len(nums)
        self.k = k
        self.tree_prod = [1] * (4 * self.n)
        self.tree_cnt = [[0] * k for _ in range(4 * self.n)]
        self._build(1, 0, self.n - 1, nums)

    def _merge(self, left_idx: int, right_idx: int, parent_idx: int):
        l_prod = self.tree_prod[left_idx]
        r_prod = self.tree_prod[right_idx]
        self.tree_prod[parent_idx] = (l_prod * r_prod) % self.k

        l_cnt = self.tree_cnt[left_idx]
        r_cnt = self.tree_cnt[right_idx]
        p_cnt = self.tree_cnt[parent_idx]

        for i in range(self.k):
            p_cnt[i] = l_cnt[i]

        for j in range(self.k):
            if r_cnt[j] > 0:
                rem = (l_prod * j) % self.k
                p_cnt[rem] += r_cnt[j]

    def _build(self, node: int, l: int, r: int, nums: List[int]):
        if l == r:
            v = nums[l] % self.k
            self.tree_prod[node] = v
            self.tree_cnt[node][v] = 1
            return
        mid = (l + r) // 2
        self._build(2 * node, l, mid, nums)
        self._build(2 * node + 1, mid + 1, r, nums)
        self._merge(2 * node, 2 * node + 1, node)

    def update(self, node: int, l: int, r: int, idx: int, val: int):
        if l == r:
            v = val % self.k
            self.tree_prod[node] = v
            for i in range(self.k):
                self.tree_cnt[node][i] = 0
            self.tree_cnt[node][v] = 1
            return
        mid = (l + r) // 2
        if idx <= mid:
            self.update(2 * node, l, mid, idx, val)
        else:
            self.update(2 * node + 1, mid + 1, r, idx, val)
        self._merge(2 * node, 2 * node + 1, node)

    def query(self, node: int, l: int, r: int, ql: int, qr: int, running_prod: int, target_x: int) -> tuple:
        ans = 0
        if ql <= l and r <= qr:
            node_cnt = self.tree_cnt[node]
            for j in range(self.k):
                if node_cnt[j] > 0:
                    if (running_prod * j) % self.k == target_x:
                        ans += node_cnt[j]
            running_prod = (running_prod * self.tree_prod[node]) % self.k
            return ans, running_prod

        mid = (l + r) // 2
        if ql <= mid:
            add_ans, running_prod = self.query(2 * node, l, mid, ql, qr, running_prod, target_x)
            ans += add_ans
        if qr > mid:
            add_ans, running_prod = self.query(2 * node + 1, mid + 1, r, ql, qr, running_prod, target_x)
            ans += add_ans

        return ans, running_prod


class Solution:
    def countKConstraintSubarrays(self, nums: List[int], k: int, queries: List[List[int]]) -> List[int]:
        seg_tree = SegmentTree(nums, k)
        n = len(nums)
        results = []

        for idx, val, start, target_x in queries:
            seg_tree.update(1, 0, n - 1, idx, val)
            ans, _ = seg_tree.query(1, 0, n - 1, start, n - 1, 1 % k, target_x)
            results.append(ans)

        return results

File: 3525-find-x-value-of-array-ii/solution.cpp
#include <vector>

using namespace std;

class Solution {
private:
    struct Node {
        int prod;
        int cnt[5];
        Node() {
            prod = 1;
            for (int i = 0; i < 5; ++i) cnt[i] = 0;
        }
    };

    int k_val;
    vector<Node> tree;

    Node mergeNodes(const Node& left, const Node& right) {
        Node res;
        res.prod = (left.prod * right.prod) % k_val;
        for (int i = 0; i < k_val; ++i) {
            res.cnt[i] = left.cnt[i];
        }
        for (int j = 0; j < k_val; ++j) {
            if (right.cnt[j] > 0) {
                int rem = (left.prod * j) % k_val;
                res.cnt[rem] += right.cnt[j];
            }
        }
        return res;
    }

    void build(int nodeIdx, int l, int r, const vector<int>& nums) {
        if (l == r) {
            int v = nums[l] % k_val;
            tree[nodeIdx].prod = v;
            tree[nodeIdx].cnt[v] = 1;
            return;
        }
        int mid = l + (r - l) / 2;
        build(2 * nodeIdx, l, mid, nums);
        build(2 * nodeIdx + 1, mid + 1, r, nums);
        tree[nodeIdx] = mergeNodes(tree[2 * nodeIdx], tree[2 * nodeIdx + 1]);
    }

    void update(int nodeIdx, int l, int r, int idx, int val) {
        if (l == r) {
            int v = val % k_val;
            tree[nodeIdx].prod = v;
            for (int i = 0; i < k_val; ++i) tree[nodeIdx].cnt[i] = 0;
            tree[nodeIdx].cnt[v] = 1;
            return;
        }
        int mid = l + (r - l) / 2;
        if (idx <= mid) {
            update(2 * nodeIdx, l, mid, idx, val);
        } else {
            update(2 * nodeIdx + 1, mid + 1, r, idx, val);
        }
        tree[nodeIdx] = mergeNodes(tree[2 * nodeIdx], tree[2 * nodeIdx + 1]);
    }

    void query(int nodeIdx, int l, int r, int ql, int qr, int& running_prod, int target_x, int& ans) {
        if (ql <= l && r <= qr) {
            for (int j = 0; j < k_val; ++j) {
                if (tree[nodeIdx].cnt[j] > 0) {
                    if ((running_prod * j) % k_val == target_x) {
                        ans += tree[nodeIdx].cnt[j];
                    }
                }
            }
            running_prod = (running_prod * tree[nodeIdx].prod) % k_val;
            return;
        }
        int mid = l + (r - l) / 2;
        if (ql <= mid) {
            query(2 * nodeIdx, l, mid, ql, qr, running_prod, target_x, ans);
        }
        if (qr > mid) {
            query(2 * nodeIdx + 1, mid + 1, r, ql, qr, running_prod, target_x, ans);
        }
    }

public:
    vector<int> countKConstraintSubarrays(vector<int>& nums, int k, vector<vector<int>>& queries) {
        // Renamed wrapper function signature if necessary, matching problem inputs
        k_val = k;
        int n = nums.size();
        tree.resize(4 * n);

        build(1, 0, n - 1, nums);

        vector<int> result;
        result.reserve(queries.size());

        for (const auto& q : queries) {
            int idx = q[0];
            int val = q[1];
            int start = q[2];
            int target_x = q[3];

            // 1. Persistent update
            update(1, 0, n - 1, idx, val);

            // 2. Query range [start, n - 1]
            int running_prod = 1 % k_val;
            int ans = 0;
            query(1, 0, n - 1, start, n - 1, running_prod, target_x, ans);

            result.push_back(ans);
        }

        return result;
    }
};